### ISSUE TITLE
Streaming

### ISSUE THREAD

**Issue by Neverlord** [actor-framework/actor-framework#531]
https://github.com/actor-framework/actor-framework/commit/f988623b7d39b776a631d00481174c0065be6287 pushed a barebone implementation of a new experimental feature in CAF: Streaming.

This issue keeps track of the implementation progress and hopefully encourages discussion with interested users and early adopters.

As of f988623b7d39b776a631d00481174c0065be6287, (scheduled) actors are now aware of *streams*. A stream is a sender-initiated contiguous flow of potentially unbound data. Streams are processed in a *pipeline*, whereas users can define such pipelines simply by composing several actors.

Currently, `scheduled_actors` have the following new member functions:
- `add_source` is used in message handlers of the form `(T) -> stream<U>`. A source generates data for downstream actors.
- `add_stage` is used in message handlers of the form `(stream<T>) -> stream<U>` to handle stream handshakes from upstream actors. A stage transforms (e.g. filters or converts) data.
- `add_sink` is used in message handlers of the form `(stream<T>) -> U` to handle stream handshakes from upstream actors. A sink terminates a stream. If `U` is not `void` then the sink *reduces* the whole stream of data into a single result.

As a motivating example, here a three example actors. One source, one stage, and one sink.

```cpp
// Models a stream source reading from a buffer.
behavior buf_source(event_based_actor* self) {
  using buf = std::deque<int>;
  return {
    [=](start_atom) -> stream<int> {
      return self->add_source(
        // initialize state
        [&](buf& xs) {
          xs = buf{1, 2, 3, 4, 5, 6, 7, 8, 9};
        },
        // get next element
        [=](buf& xs, downstream<int>& out, size_t num) {
          auto n = std::min(num, xs.size());
          for (size_t i = 0; i < n; ++i)
            out.push(xs[i]);
          xs.erase(xs.begin(), xs.begin() + static_cast<ptrdiff_t>(n));
        },
        // check whether we reached the end
        [=](const buf& xs) {
          return xs.empty();
        }
      );
    }
  };
}

// Models a stream stage that filters even numbers.
behavior filter_even(event_based_actor* self) {
  return {
    [=](stream<int>& in) -> stream<int> {
      return self->add_stage(
        // input stream
        in,
        // initialize state
        [=](unit_t&) {
          // nop
        },
        // processing step
        [=](unit_t&, downstream<int>& out, int x) {
          if (x & 0x01)
            out.push(x);
        },
        // cleanup
        [=](unit_t&) {
          // nop
        }
      );
    }
  };
}

// Models a stream sink that sums up all numbers in a stream.
behavior sum_up(event_based_actor* self) {
  return {
    [=](stream<int>& in) {
      return self->add_sink(
        // input stream
        in,
        // initialize state
        [](int& x) {
          x = 0;
        },
        // processing step
        [](int& x, int y) {
          x += y;
        },
        // cleanup and produce result message
        [](int& x) -> int {
          return x;
        }
      );
    }
  };
}

void caf_main(actor_system& sys) {
  scoped_actor self{sys};
  auto source = sys.spawn(buf_source);
  auto stage = sys.spawn(filter_even);
  auto sink = sys.spawn(sum_up);
  // read: self -after- (sink -after- (stage -after- source))
  //   ie: data will flow from souce to sink and the result is received by self
  auto pipeline = self * sink * stage * source;
  self->send(pipeline, start_atom::value);
  self->receive([](int x) {
    cout << "the result of the stream is: " << x << endl;
  });
}
```

Note that `stream<T>` is a *marker*. Users can never send a `stream<T>` object directly from one actor to another.

I encourage all users interested in this feature to play around with it and give feedback. Both in terms of shortcomings of the current implementation as well as suggestions for future extensions or additional APIs.

An explicit goal is to have a new declarative API for doing data processing in cluster environments (via a new `cluster` API in CAF).

The following list documents progress on the new streaming API.

- [x] Provide sophisticated, user-facing API
   + [x] `self->stream(pipeline, ...)` for starting an "ad-hoc stream"
   + [x] Expose internals (to a certain degree) to allow full control over up- and downpaths
   + [x] Allow user-defined up- and downstream policies
- [x] Launch as experimental feature

**Comment by CyberBrainX** [comment]
Nice, especially the cluster module!
Will it be possible to define different types of pipelines? Like a single source and multiple sinks (load balancing) or multiple sources and a single sink (collecting) or even multiple sources to multiple sinks? Or need this be done in a custom stage?

**Comment by Neverlord** [comment]
> Will it be possible to define different types of pipelines?

Load balancing and collecting will be supported. Actors can also access the streams they participate in and add more sinks or sources at runtime. Fine-tuning can be done with policies. For example, a policy handles how multiple sinks are handled. Examples are broadcasting all messages or load balance between the sinks.

**Comment by Neverlord** [comment]
I moved integration of streams into the network layer to #571, clustering to #572, and a streamlined high-level API to #573.

This feature is already large enough as it stands and should be shipped as minimal experimental feature sooner rather than later. I think it's better to ship the currently available subset of planned features with 0.15.4.
